Data settings addition

rahulKQL opened the pull request:
This commits contains: 
- A class with static method for conversion of `BigtableOptions` to `BigtableDataSettings`.
- Basic test cases for the same, more to add once finalised.

**Note:** This work is still in progress, opening PR for review purpose.

rahulKQL commented:
@igorbernstein2  Please review

rahulKQL commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,174 @@
+      batchingSettingsBuilder.setDelayThreshold(Duration.ofMillis(autoFlushMs));
+    }
+
+    builder.bulkMutationsSettings().setBatchingSettings(batchingSettingsBuilder.build());
```
Will add more mutationsSettings methods

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,174 @@
+      builder.bulkMutationsSettings().setRetrySettings(retrySettings);
+      builder.checkAndMutateRowSettings().setRetrySettings(retrySettings);
+      builder.readModifyWriteRowSettings().setRetrySettings(retrySettings);
+      builder.readRowsSettings().setRetrySettings(retrySettings);
```
* readRows should use the long timeout from CallOptionsConfig for totalTimeout
* readRows should use the ReadPartialRowTimeoutMillis as the rpc timeout

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,174 @@
+      builder.mutateRowSettings().setRetrySettings(retrySettings);
+      builder.sampleRowKeysSettings().setRetrySettings(retrySettings);
+    } else {
+
```
Should retries be disabled for other methods as well?

rahulKQL replied:
Initially I have disabled retries on all settings(). Would by default enable retries for all, and then for user specified settings we can disable retries. 
does that sounds good?

igorbernstein2 replied:
I think so. Also, please be careful: not all methods are idempotent. For example CheckAndMutate & ReadModifyWrite are not

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,174 @@
+    // userAgent can be set on headers.
+    String userAgent = options.getUserAgent();
+    if (null != userAgent)
+      builder.setHeaderProvider(FixedHeaderProvider.create("USER_AGENT", userAgent));
```
The header name is wrong here. Please see the logic in:
https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcHeaderInterceptor.java#L68

rahulKQL replied:
I have added two sets of header in new commit as It could be set in headerProvider as well as ManagedChannelBuilder#userAgent.

sduskis replied:
This actually won't work.  gRPC sets overrides the user agent based on configuration that happens at the creation of the Channel.  `WithUserAgent()` is likely the only way to implement this feature.

rahulKQL commented on bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,72 @@
+package com.google.cloud.bigtable.hbase;
```
Sorry for late response, Would add those in today.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,174 @@
+
+    builder.setTransportChannelProvider(InstantiatingGrpcChannelProvider.newBuilder()
+        .setChannelsPerCpu(options.getChannelCount())
+        .setMaxInboundMessageSize(retryOptions.getStreamingBufferSize()).build());
```
streamingBufferSize is a different concept from inboundMessageSize:
inboundMessageSize is the size of the grpc's internal buffer
streamingBufferSize doesnt appear to be used, but was intended to be used by RowMerger to limit the number of rows to get merged.

I think streamingBufferSize should be removed altogether

rahulKQL commented:
#### In this commit:
- individual settings method, with retry check.
- default method to set RetrySettings.
- added/fixed code based on suggestions.
- added many inline comment on doubtful points.
- Added default RPC timeout as 100ms in case of No Retry

Would add test cases & removed unneeded comments once it gets reviewed.
@igorbernstein2 Could you please have a look at it

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+    }
+    long maxIn = bulkOptions.getMaxInflightRpcs();
+    if(maxIn > 0) {
+      flowControlBuilder.setMaxOutstandingElementCount(maxIn);
```
This should be maxInflightRpcs * bulkMaxRowKeyCount

rahulKQL replied:
oh, will update the same.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+
+  // NOTE: Included Code.UNAUTHENTICATED for default retryCodes
+  private static final Set<Code> DEFAULT_RETRY_CODES =
+      ImmutableSet.of(Code.DEADLINE_EXCEEDED, Code.UNAVAILABLE, Code.ABORTED, Code.UNAUTHENTICATED);
```
I dont think we should change the default veneer retry codes

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+      ImmutableSet.of(Code.DEADLINE_EXCEEDED, Code.UNAVAILABLE, Code.ABORTED, Code.UNAUTHENTICATED);
+
+  // NOTE: set rpcTimeout 100ms, without retries.
+  private static final Duration DEFAULT_RPC_TIMEOUT = Duration.ofMillis(100);
```
I think this should come from calloptionsconfig

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+    } else {
+      builder.sampleRowKeysSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
```
doesn't setSimpleTimeoutNoRetries already clear the codes?

rahulKQL replied:
Yes, Will remove unnecessary `getRetryableCodes().clear()`

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.readRowsSettings()
+          .setIdleTimeout(ofMillis(options.getRetryOptions().getReadPartialRowTimeoutMillis()))
```
This is incorrect idleTimeout is the delay for the user leaving a stream open not the timeout for the server. You want to set the streamingWaitTimeout. There are 2 ways to do it:
1. the rpcTimeout of the RetrySettings gets translated into a waitTimeout
2. set it in calloptions

1 will work for when retries are enabled. However when retries are disabled, you would have to augment the calloptions at the callsite, which is awkward. Please open a ticket in gax-java to expose waitTimeout at the callable level

rahulKQL replied:
I would open an issue stating this above scenario. Thanks

rahulKQL commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,310 @@
+    // configurations for RPC timeouts
+    retryBuilder
+        .setInitialRpcTimeout(ofMillis(options.getCallOptionsConfig().getShortRpcTimeoutMs()))
+        .setMaxRpcTimeout(ofMillis(retryOptions.getReadPartialRowTimeoutMillis()))
```
@igorbernstein2 
I set the `rpcTimeout` with `readPartialTimoutMillis`.. As we are expecting clients must have retry enabled then, would this be fine?

igorbernstein2 replied:
This is mixing up different concepts. For readRows you want to set
initialRpcTimeout = maxRpcTimeout = readPartialRowTimeoutMillis

For non-streaming methods you want to set 
initialRpcTimeout = maxRpcTimeout = shortRpcTimeout

I'm guessing the easiest way to get there is have readRows not use this method

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+      throws IOException, GeneralSecurityException {
+    if (!options.getRetryOptions().enableRetries()) {
+      throw new IllegalStateException(
+          "Retry is must for BigtableDataSettings configuration from BigtableOptions.");
```
The message is a bit awkward. Maybe something like "Disabling retries is not currently supported"

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+    builder.setAppProfileId(options.getAppProfileId());
+
+    LOG.debug("endpoint host %s.", options.getDataHost());
+    LOG.debug("endpoint host %s.", options.getPort());
```
This should be logged elsewhere

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+
+    buildCredentialProvider(builder, options.getCredentialOptions());
+
+    buildBulkOptions(builder, options);
```
The naming is a bit inconsistent here, please name it after the method name like `buildBulkMutations`

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+    buildSampleRowKeys(builder, options);
+
+    // TODO: would it map to GrpcHeaderInterceptor? or we should build userAgent
+    // using ManagedChannelBuilder AND
```
If the user agent on the Managed channel gets preserved, I think that would be cleaner. If it doesn't, add note that it doesnt

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+
+    // TODO: implementation for channelCount or channelPerCPU
+    ManagedChannelBuilder channelBuilder = ManagedChannelBuilder
+        .forAddress(options.getDataHost(), options.getPort())//
```
stray //

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+    // TODO: implementation for channelCount or channelPerCPU
+    ManagedChannelBuilder channelBuilder = ManagedChannelBuilder
+        .forAddress(options.getDataHost(), options.getPort())//
+        .userAgent(options.getUserAgent());
```
Please verify if this works and remove the other approach if it does

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+    }
+    if (maxInflightRpcs > 0) {
+      flowControlBuilder.setMaxOutstandingElementCount(maxInflightRpcs * bulkMaxRowKeyCount);
+    }
```
Please group flowController settings together

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param bulkMutation a {@link BulkOptions} object.
+   */
+  private static RetrySettings defaultRetrySettings(BigtableOptions options) {
```
Please rename this to something more appropriate. This method doesn't represent default settings. Maybe something like `buildIdempotentRetrySettings`?

igorbernstein2 commented on bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java:
```diff
@@ -0,0 +1,265 @@
+        .setMaxRpcTimeout(ofMillis(retryOptions.getReadPartialRowTimeoutMillis()))
+        .setTotalTimeout(ofMillis(options.getCallOptionsConfig().getLongRpcTimeoutMs()));
+
+    // TODO: an option to set RetryOptions#allowRetriesWithoutTimestamp
```
I dont think this will ever be supported as a setting. Please throw an unsupported exception instructing users to use ajay's unsafe Mutation factory method

rahulKQL commented:
## Changes

#### BigtableVaneerSettingsFactory
 - Added UserAgent through `HeaderProvider` with `GrpcUtil.USER_AGENT_KEY` which internally translated to ManagedChannel [here](https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java#L228).
 - Moved to [InstantiatingGrpcChannelProvider](https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java#L65) for ChannelProvider, which gives option for Headers along with channel PoolSize.
 - Overriding non-idempotent RPCs only when SHORT_RPC_TIMEOUT is other than default value(i.e. 60seconds).
 - For **BigtableTableAdminSettings**: overriding Endpoint, CredentialProvider & TransportChannelProvider configuration only.


#### TestBigtableVaneerSettingsFactory
- Added [test case](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-4acee08860ebdfaf1d83ea5d747b5508R138) which tests `BigtableDataClient` as well `BigtableTableAdminClient` with an actual Bigtable, This test case runs only when it founds "**_test.client.project.id_**" & "**_test.client.project.id_**" VM arguments.



#### Question:
My purpose for using InstantiatingGrpcChannelProvider is to be able to create channelPools, however 
according to [this comment](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973#discussion_r235973614), it seems ManagedChannel#WithUserAgnet() should be used to setting user-agent, So for considering multipleChannel:

Should we use the [BigtableSession#createChannelConfig](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java#L478) as it takes care of creating ChannelPool?

rahulKQL commented:
@sduskis, @igorbernstein2 

Please have a look at this and let me know of your thought.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+
+    buildSampleRowKeysSettings(builder, options);
+
+    String userAgent = BigtableVersionInfo.CORE_UESR_AGENT + "," + options.getUserAgent();
```
s/CORE_UESR_AGENT/CORE_USER_AGENT

rahulKQL replied:
Renamed the constant

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+    buildSampleRowKeysSettings(builder, options);
+
+    String userAgent = BigtableVersionInfo.CORE_UESR_AGENT + "," + options.getUserAgent();
+    HeaderProvider headers = FixedHeaderProvider.create(GrpcUtil.USER_AGENT_KEY.name(), userAgent);
```
Does this actually work or does grpc clobber it?

rahulKQL replied:
I think It is getting translated to both the clients. 

I have added a simple test case for BigtableDataSettigns & BigtbaleTableAdminSettings with [ActualBigtable here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-4acee08860ebdfaf1d83ea5d747b5508R140) (which we may remove as it might not needed here). 

So It created 32 ManagedChannels under channels pool when I run it in my local machine, Also It had the persisted user-agent on those channels. (attaching snapshot from debug console).
![Uploading DataClient-user-agent-verification.png…](BigtableDataClient-User-Agent)

igorbernstein2 replied:
I think this needs to be verified on the server side and there should be a test to confirm this.  Please add a todo

igorbernstein2 replied:
Can you add a comment about how gax-java has special handling for  the USER_AGENT_KEY:
https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcHeaderInterceptor.java#L63-L74

rahulKQL replied:
Added a separate test case for User-agent [TestUserAgent](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-8b4fb9399d066a371cb9b54a9a6d871cR59), verifying header with the help of HeaderServerInterceptor for [PlainText](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-8b4fb9399d066a371cb9b54a9a6d871cR91) as well as [TLS](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-8b4fb9399d066a371cb9b54a9a6d871cR122) Negotiation (_Using Client_trust, Server_Cert & Private key for the TLS user agent test case_).

Added [an inline comment here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R283) about User-Agent handling.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+
+    BigtableTableAdminSettings.Builder adminBuilder = BigtableTableAdminSettings.newBuilder();
+
+    adminBuilder.setInstanceName(com.google.bigtable.admin.v2.InstanceName
```
Please add a note to stop using typesafe names

rahulKQL replied:
Updated classes to use with raw ids

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildMutateRowSettings(Builder builder, BigtableOptions options) {
+    builder.mutateRowSettings()
```
where are the retry codes configured? (here and everywhere else)

rahulKQL replied:
As of now, not overriding `RetryCodes` of `BigtableDataSettings`. 
Defaults retrycode in GCJ:
**[DEADLINE_EXCEEDED, UNAVAILABLE]** --> sampleRowKeysSettings, readRowsSettings, mutateRowSettings.
**empty** for bulkMutationsSettings, readModifyWriteRowSettings, checkAndMutateRowSettings.

Default retrycodes in [bigtable-client](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/RetryOptions.java#L50).

However, I can override these with operation specific retry codes.

igorbernstein2 replied:
There are few more defaults for gcj in the overlay. The gapic generated client settings can't express all of the settings, so they are overlaid here:
https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java#L273-L321

This is where bulkMutations have their defaults set.

I'm trying to understand why we shouldn't copy the retry codes from cbt?

rahulKQL replied:
I think for **Non-idempotent** operation we can utilize GCJ's default values, as we are overriding it with shortRpcTimeout & no retryCodes (_here both adapters have the default value of 1 min_).

In case of **Idempotent** operation default values varies alot & 
 - [CBT has retryCode](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/RetryOptions.java#L49-L54) as `{DEADLINE_EXCEEDED, UNAVAILABLE, ABORTED, UNAUTHENTICATED}` 
- [GCJ's retryCodes](https://github.com/googleapis/googleapis/blob/master/google/bigtable/v2/bigtable_gapic.yaml#L35-L39) are only `{DEADLINE_EXCEEDED, UNAVAILABLE}`.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+
+    builder.readRowsSettings()
+        .setRetrySettings(retryBuilder.build());
+  }
```
Please update google-cloud-java dep and add config for single row reads

rahulKQL replied:
I think we would need readRowSettings() in the [BigtableDataSettings](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java) as well. For now, It is available in [EnhanccedBigtableStubSettings](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java) only.

Let me know If I should add this change.

igorbernstein2 replied:
Please do, I apologize for the omission

rahulKQL replied:
Added [buildReadRowSettings()](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R174) for point reads.

As of now set it using the same configuration as readRowsSettings.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+        credentialsProvider = BigtableStubSettings.defaultCredentialsProviderBuilder().build();
+        break;
+      case P12:
+      case SuppliedCredentials:
```
should this be something along the lines of 
```
credential = ((UserSuppliedCredentialOptions)credentialOptions).getCredential();
credentialsProvider = FixedCredentialsProvider.create(credential);
```

rahulKQL replied:
Updated [buildCredentialProvider()](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-3b1db200ad0c45bb25cfa85e47b21766R284) to use [CredentialFactory#getCredentials](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/CredentialFactory.java#L129-L156)

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,312 @@
+      case DefaultCredentials:
+        credentialsProvider = BigtableStubSettings.defaultCredentialsProviderBuilder().build();
+        break;
+      case P12:
```
Shouldnt this do something similar to 
https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/CredentialFactory.java#L169-L181

rahulKQL replied:
Thanks for the pointer, Have updated this class to utilize the same.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java:
```diff
@@ -0,0 +1,298 @@
+    adminBuilder.stubSettings()
+        .setCredentialsProvider(buildCredentialProvider(options.getCredentialOptions()));
+
+    String userAgent = BigtableVersionInfo.CORE_USER_AGENT + "," + options.getUserAgent();
```
Please add something here that would allow us to disambiguate when the hbase client is using the gcj adapter

rahulKQL replied:
Does ["VENEER_ADAPTER"](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R54) sound good?

rahulKQL commented:
## Summary of Client Settings changes

After many Iteration now in this PR, as of now we are:
  - Overriding both client setting's endpoint with BigtableOptions endpoint [here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R83).
  - Utilizing existing [CredentialFactory#getCredentials](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R138) for CredentialProvider.
  - Providing ["VENEER_ADAPTER"](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R300) in the User-Agent to be able to distinguish GCJ adapters.
  - Created TransportChannelProvider using InstantiatingGrpcChannelProvider [here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R306):
    - Kept ChannelsPerCpu to 2 (Though it BigtableOptions has 4 ChannelsPerCpu as default).
    - Kept MaxInboundMessageSize to CBT's default(i.e. 256MB).
    - For plainText Negotiation using [ChannelConfigurator](https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java#L450-L462) (_would this be ok?, as its not stable yet?_).
  - Created [buildIdempotentRetrySettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L260-L284) to configure generic RetrySettings.

    ### [createBigtableDataSettings](BigtableVeneerSettingsFactory)
      - As of now we are not supporting disabling the retries.
      - In case of [bulkMutationsSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L150-L184), We override:
            - BatchingSettings with BulkOptions values.
            - RetrySettings with idempotentRetrySettings.
            - RetryCodes with BigtableOptions's Retries Code.
      - In case of [sampleRowKeysSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L208-L216), [mutateRowSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L218-L226) and [readRowSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L228-L236), We overrides
             - RetrySettings with idempotentRetrySettings.
             - RetryCodes with BigtableOptions's Retries Code.
      - In case of [readRowsSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L238-L258), We override
             - RetrySettings with idempotentRetrySettings considering streaming method according to this suggestion.
             - RetryCodes with BigtableOptions's Retries Code.
      - In case of checkAndMutateRowSettings & readModifyWriteRowSettings, We override SimpleTimeoutNoRetries when shortRpcTimeoutMs is not default.
       
    ### [createTableAdminSettings](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L113-L132)
      - Only Overrided [endpoint, CredentialsProvider & TransportChannelProvider](https://github.com/rahulKQL/cloud-bigtable-client/blob/dataSettings_addition/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java#L123-L129). Do we need to override any specific settings for this?

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+    //ChannelBuilder with provided value.
+    HeaderProvider headers = FixedHeaderProvider.create(USER_AGENT_KEY.name(), userAgent);
+
+    InstantiatingGrpcChannelProvider.Builder builder =
```
Please a TODO to refactor google-cloud-java to expose a static `defaultTransportChannelProvider` method and then customize it.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+            .setHeaderProvider(headers)
+            .setEndpoint(endpoint)
+            .setPoolSize(options.getChannelCount())
+            .setMaxInboundMessageSize(MAX_MESSAGE_SIZE);
```
This seems to ignore BigtableOptions.getDataChannelCount(). Might be better to remove the per cpu channel counts and just use the absolute channel count via `setPoolSize`

rahulKQL replied:
Removed ChannelPerCPU() config.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+  public static BigtableDataSettings createBigtableDataSettings(@Nonnull final BigtableOptions options) throws IOException {
+    checkState(options.getProjectId() != null, "Project ID is required");
+    checkState(options.getInstanceId() != null, "Instance ID is required");
+    checkState(options.getRetryOptions().enableRetries(), "Disabling retries is not currently supported.");
```
I don't think useCachedDataPool option is supported here, please throw an error if its enabled

rahulKQL replied:
Added state check for useCachedChannel:
```java
checkState(!options.useCachedChannel(), "cachedDataPool is not currently supported.");
````

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+  public static BigtableDataSettings createBigtableDataSettings(@Nonnull final BigtableOptions options) throws IOException {
+    checkState(options.getProjectId() != null, "Project ID is required");
+    checkState(options.getInstanceId() != null, "Instance ID is required");
+    checkState(options.getRetryOptions().enableRetries(), "Disabling retries is not currently supported.");
```
useEmulator option is not covered here. Please throw an error and add a todo to support it

rahulKQL replied:
I hope checking against Environment variable would be sufficient:
```
String emulatorHost = System.getenv(BIGTABLE_EMULATOR_HOST_ENV_VAR);
checkState(emulatorHost == null, "Emulator Hosting is not supported yet.");
````

igorbernstein2 replied:
Actually nevermind, the emulator should work fine. Sorry for the noise

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+    InstantiatingGrpcChannelProvider.Builder builder =
+        InstantiatingGrpcChannelProvider.newBuilder()
+            .setChannelsPerCpu(2)
+            .setHeaderProvider(headers)
```
Shouldn't the header provider be set on the top level settings? If I understand correctly, the header provider on the channel provider is set by the client context during build: 
https://github.com/googleapis/gax-java/blob/122eda65b4857cce676b53f07d19fef59dc0cb15/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L151

rahulKQL replied:
Unfortunately, while [creating `BigtableStubSettings` in `EnhancedBigtableStub#create`](https://github.com/rahulKQL/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java#L88-L94), we are not considering client’s `HeaderProvider` value present in the argument.

I cross checked,  only default header i.e. `x-goog-api-client` is available at ClientContext’s Xtor. We would need user-agent to be available at `TransportChannelProvider`, for that reason I created `InstantiatingGrpcChannelProvider`.
Also we had to set `ManagedChannelBuilder#usePlainText` to support plainText negotiation rpcs.

Do let me know If I should raise PR to consider HeaderProvider at `EnhancedBigtableStub#create` before ClientContext creation.

igorbernstein2 replied:
Please do, its a bug: the header provider should be copied over in the enhanced stub

rahulKQL replied:
[PR#4491](https://github.com/googleapis/google-cloud-java/pull/4491) for headerProvider addition is merged now. 

Once the next release(i.e. v0.81.0) is available, I will have this PR rebase and updated.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+    retryBuilder
+        .setInitialRpcTimeout(shortRpcTimeout)
+        .setMaxRpcTimeout(shortRpcTimeout)
+        .setTotalTimeout(ofMillis(options.getCallOptionsConfig().getLongRpcTimeoutMs()));
```
This doesn't look right. I think that for all rpcs except ReadRows (plural), the mapping should be:
- if useTimeouts == true, then initialRpcTimeout = maxRpcTimeout = shortRpcTimeout
- if useTimeouts == false, then initialRpcTimeout = maxRpcTimeout = [6 minutes?](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java#L74)
- totalTimeout = RetryOptions.getMaxElapsedBackoffMillis
- maxAttempts = 0 (disabled) (@sduskis, am I correct that classic java client doesn't limit attempts for other rpcs?)

For readRows
- initialRpcTimeout = maxRpcTimeout = RetryOptions.readPartialRowTimeoutMillis
- if (useTimeouts) totalTimeout = longRpcTimeoutMs
- maxAttempts = RetryOptions.getMaxScanTimeoutRetries

rahulKQL replied:
Thanks a lot for detailed values. Have followed the same for configuring all idempotent retry settings.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+    retryBuilder
+        .setInitialRetryDelay(ofMillis(retryOptions.getInitialBackoffMillis()))
+        .setRetryDelayMultiplier(retryOptions.getBackoffMultiplier())
+        .setMaxRetryDelay(ofMillis(retryOptions.getMaxElapsedBackoffMillis()))
```
getMaxElapsedBackoffMillis is the overall timeout, not the max delay. It seems like the classic client hardcodes maxRetryDelay to 1 minute:
https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java#L338

rahulKQL replied:
Thanks for the reference, Have set MaxRetryDelay to 1 min using `MAX_RETRY_TIMEOUT_MS = 60_000`

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+        .setInitialRetryDelay(ofMillis(retryOptions.getInitialBackoffMillis()))
+        .setRetryDelayMultiplier(retryOptions.getBackoffMultiplier())
+        .setMaxRetryDelay(ofMillis(retryOptions.getMaxElapsedBackoffMillis()))
+        .setMaxAttempts(retryOptions.getMaxScanTimeoutRetries());
```
There is no max attempts in the classic client. Please use 0 to disable

rahulKQL replied:
removed overriding of maxAttempts(except ReadRows), it would now be defaults to `setMaxAttempts(0)`.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+    RetrySettings.Builder retryBuilder = builder.readRowsSettings().getRetrySettings().toBuilder()
+        .setInitialRetryDelay(ofMillis(retryOptions.getInitialBackoffMillis()))
+        .setRetryDelayMultiplier(retryOptions.getBackoffMultiplier())
+        .setMaxRetryDelay(ofMillis(retryOptions.getMaxElapsedBackoffMillis()))
```
maxRetryDelay = 1 minute

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+   * than 60_000 ms.
+   */
+  private static void buildReadModifyWriteSettings(Builder builder, long rpcTimeoutMs) {
+    if(rpcTimeoutMs != SHORT_TIMEOUT_MS_DEFAULT) {
```
why if? (here and above)

rahulKQL replied:
If shortTimeOutMs is at default then assuming the client did not customize BigtableOption's settings, 
that is why we are not overriding the underlying settings here(or directly using defaults of GCJ's settings).

igorbernstein2 replied:
I think you should override with the BigtableOptions in case the diverge

rahulKQL replied:
> override with the BigtableOptions in case the diverge
By saying above comment,
do you mean we should check all configuration against their default value(present in BigtableOption) & then override only those which are different?
OR
Should we override only in case of specific operation(say non-streming operations) by checking BigtableOption's present value against BigtableDataSetting's default value?

igorbernstein2 replied:
I just meant you should remove the if statement and always setSimpleTimeoutNoRetries disrarding if the if its the default value or not. If we decide that to change the gapic config to 30 secs, but forget to change SHORT_TIMEOUT_MS_DEFAULT we will have confusing behavior: the user will check their BigtableOptions and see SHORT_TIMEOUT_MS_DEFAULT = 1 minute, but their rpcs will exceed deadline @ 30secs

rahulKQL replied:
Thanks for the clarification!! 
I followed the same [here](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/1973/files#diff-6a492166d0005302ce3c542cdcf00ad5R96).

rahulKQL commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,327 @@
+   */
+  public static BigtableDataSettings createBigtableDataSettings(@Nonnull final BigtableOptions options) throws IOException {
+    checkState(options.getProjectId() != null, "Project ID is required");
+    checkState(options.getInstanceId() != null, "Instance ID is required");
```
I have added these for an extra layer of validation. Would remove them.

igorbernstein2 commented:
I just submitted https://github.com/googleapis/google-cloud-java/pull/4507, it should help with the transport builder

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,314 @@
+
+  //Identifier to distinguish between CBT or GCJ adapter.
+  private static final String VENEER_ADAPTER =  BigtableVersionInfo.CORE_USER_AGENT+","
+      + "VENEER_ADAPTER,";
```
nit: Why is this capitalized?

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,314 @@
+    checkState(options.getRetryOptions().enableRetries(), "Disabling retries is not currently supported.");
+    checkState(!options.useCachedChannel(), "cachedDataPool is not currently supported.");
+
+    //TODO:add configuration for emulator hosting.
```
what is the todo for?

rahulKQL replied:
This was intended to do separate emulator configuration, but I think we agreed on not to have the separate configuration.

igorbernstein2 commented on bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVeneerSettingsFactory.java:
```diff
@@ -0,0 +1,314 @@
+    RetrySettings.Builder retryBuilder = retrySettings.toBuilder();
+
+    if (retryOptions.allowRetriesWithoutTimestamp()) {
+      LOG.warn("Retries without Timestamp does not support yet.");
```
This should throw instead of warn

rahulKQL commented:
@sduskis 

Will be just rebasing and updating these three last comments
